### PR TITLE
Add synthetic constructors only to Java annotations

### PR DESCRIPTION
**Pull request by Jasper-M**
Backport of #9981, but with improvements to be forward-ported.

### CODE REVIEW

**Comment by Jasper-M** [comment]
Is there some kind of 2.12 alternative for this `javaVersion: 9+` thing? 
Also, does anyone understand why the more conservative change resulted in [this](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main/4973/console) NullPointerException?

**Comment by som-snytt** [comment]
I did take a look earlier and wanted to :smiley: on "go nuts". The unfortunate build error dissuaded me from looking further. I almost commented, "Just when you think it's an easy backport..."

**Comment by som-snytt** [comment]
I'm required to do further penance because I broke the community build.

Is this considered a desirable backport because people on 2.12 are likely to be using `-release` or whatever the option is currently?

Or was it hoped to be fruit of a low-hanging variety. Like strawberries.

IIRC retronym did work on both branches: it's genuine retronym, not a knock-off.

I saw the NPE in test but hadn't looked yet.

**Comment by Jasper-M** [comment]
I use it on 2.12. I don't know about other people.

**Comment by som-snytt** [comment]
Say no more. I too don't know about other people, they're so weird.

**Comment by som-snytt** [comment]
Lukas's javaVersion was an easier backport so I did that first, at your urging. POC
```
-- 1 - neg/t12565.scala                          [skipped on Java 17, only running on 19+]
```
https://github.com/scala/scala/pull/10021

**Comment by som-snytt** [comment]
The zinc NPE is an interaction with Lukas's https://github.com/sbt/zinc/commit/c25d95665406f0c51b1b3b9b239566765c2b1fcf where the common elements is lazy completion of annotation info, but I don't understand the cogs yet. That is really the point of the whole misadventure.

**Comment by SethTisue** [comment]
@Jasper-M the window for 2.12.16 is closing soon

**Comment by lrytz** [comment]
I can explain the NPE, a bunch of things playing together here...

```scala
  // current 2.13 version
  (sflags & (INTERFACE|JAVA_ANNOTATION)) == (INTERFACE|JAVA_ANNOTATION)
```

Note that we're using `sflags`, which is defined as `val sflags = jflags.toScalaFlags`. Interestingly:

```
// 2.12
scala> typeOf[Deprecated].typeSymbol.debugFlagString
res0: String = <java> <annotation>

// 2.13
scala> typeOf[Deprecated].typeSymbol.debugFlagString
val res0: String = <interface> abstract <java> <defaultparam/trait> <annotation>
```

which is due to this (really nice) cleanup: https://github.com/scala/scala/commit/7c9e3724ee#diff-b47f6177947626919b42f078a149a0e822e464c4a96b39c2dfbaef1f2e43ea34

So backporting the 2.13 version to 2.12 meant that no more artificial constructors were added to Java annotations at all (`INTERFACE` is not there on 2.12). This ended up in an NPE (instead of a better error message) because in `typedSelect`:

https://github.com/scala/scala/blob/29e696fbc34818feb2bee4f36be4266611dd941b/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5069
https://github.com/scala/scala/blob/29e696fbc34818feb2bee4f36be4266611dd941b/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5048-L5051

The `qual` here is `new Deprecated`, the selected name is `<init>`. The `qual.symbol` is null. The correct symbol here would be `qual.tpe.typeSymbol`.

I'm all for the nutty version, also for forward-porting it to 2.13.

**Comment by Jasper-M** [comment]
Thanks @lrytz and @som-snytt. I'm on holiday so I had more or less resigned to missing 2.12.16.

**Review by som-snytt — APPROVED** [review]
It's no coincidence that the dashboard icon on my car for "lane keep assistance", that keeps you from driving off a cliff, is `LKAS`.

I'll offer the forward port.

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 1]
```diff
@@ -606,8 +600,6 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
     val sflags = jflags.toScalaFlags
     if (jflags.isPrivate) {
       val isConstructor = pool.getName(u2).value == "<init>" // opt avoid interning a Name for private methods we're about to discard
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: Some unusedness.

**Comment by som-snytt** [comment]
The diff is an ideal `+27 -27`.

**Comment by SethTisue** [comment]
@lrytz you already approved the forward-port; did you mean to approve this one as well?

**Comment by lrytz** [comment]
Hard to say who should approve it :-) LGTsom & LGTM, so 🚀

**Pull request merged**